crates/subspace-farmer/src/farmer_cache.rs

```diff
@@ -153,12 +153,58 @@ where
             }
         }
     }
+
+    fn has_free_capatity(&self) -> bool {
+        if !self.dangling_free_offsets.is_empty() {
+            return true;
+        }
+
+        self.backends.iter().any(|backend| backend.free_size() > 0)
+    }
+
+    fn should_include_key(&self, peer_id: PeerId, key: PieceIndex) -> bool {
+        let key = DistanceForKey::new(peer_id, key.to_multihash());
+        self.should_include_key_internal(key)
+    }
+
+    fn should_include_key_internal(&self, key: DistanceForKey) -> bool {
+        if self.stored_pieces.contains_key(&key) {
+            return false;
+        }
+
+        if self.has_free_capatity() {
+            return true;
+        }
+
+        let top_key = self.stored_pieces.last_key_value().map(|(key, _)| key);
+
+        if let Some(top_key) = top_key {
+            *top_key > key
+        } else {
+            false // TODO: consider adding error here
+        }
+    }
+
+    fn insert(
+        &mut self,
+        key: DistanceForKey,
```

Comment:
Because here is just checking to see if I need to replace old oiece with new one in stored _offsets, and pop the old one out. I thought I'd rename this fn to should_repalce.

Reply:
Hm... it is a confusing name. Insert that doesn't insert (but rather removes stuff) is not a good API.

While the suggestion to not use `DistanceForKey` still stands, I think I now would like to see a separate PR where the whole `PieceCachesState` is moved into submodule with all properties private so we can clearly see what APIs are called. That way we don't manually set properties from various places and don't make methods like `should_include_key_internal` available. Then this PR can be rebased and will look cleaner and easier to verify correctness as opposed to using methods in some cases and editing properties directly in others.

Reply:
> While the suggestion to not use DistanceForKey still stands, I think I now would like to see a separate PR where the whole PieceCachesState is moved into submodule with all properties private so we can clearly see what APIs are called. That way we don't manually set properties from various places and don't make methods like should_include_key_internal available. Then this PR can be rebased and will look cleaner and easier to verify correctness as opposed to using methods in some cases and editing properties directly in others.

I agree, let me deal with this first, it will be clearer code.